[#8] feat: 생성자, 생성일, 중요도 별 filtering 기능 구현

imhjlov opened the pull request:
## 수정 사항 간략한 한줄 요약
- 생성자, 생성일, 중요도 별 filtering 기능을 구현 하였습니다.

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/30456236/131013723-270357f8-ed68-4ca8-9f71-41334922ef4c.gif)

## 수정 사항들 자세한 내용
- dropdown에서 생성자, 생성일, 중요도를 선택할 수 있습니다.
- input창에 입력 시 dropdown에서 선택 된 filter 기준에 따라 목록이 보여 집니다.
- input 창에 입력 할 때마다 filter됩니다.

## 기타 질문 및 특이 사항
- localstorage에서 데이터를 불러오게 하였는데 다른 좋은 방법이 없는지 궁금합니다.
- 전역에 저장된 todos를 이용하면 무한 랜더링이 됩니다.
- 💡 버그가 있으면 꼭 알려주세요.
- ✅ merge하는 과정에서 drag and drop이 제 기능을 하지 못 합니다. pull 받고 한번 봐주세요.

## 체크 리스트 
- [x] 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 팀원들 모두 해당 PR에 대하여 알고 있습니다.

imhjlov committed:
[#8] feat: 생성자, 생성일, 중요도 별 filtering 기능 구현

Co-authored-by: dongwonnn <dongwonnn@users.noreply.github.com>

imhjlov committed:
[#23] 텍스트를 입력하지 않을 때 state 값이 변경 되지 않는 버그 수정

imhjlov committed:
Revert "[#23] 텍스트를 입력하지 않을 때 state 값이 변경 되지 않는 버그 수정"

This reverts commit 1cd2b056e50f285f553e1da14b55b2622d0e5c5e.

blingblin-g commented on src/components/filter/FilterInput.tsx:
```diff
@@ -1,11 +1,41 @@
+    } else if (originalTodos){
+      dispatch({
+        type: 'SAVE',
+        saveTodo:originalTodos
```
filterdData 또는 originalTodos를 받아서 {
type: 'SAVE',
saveTodo: 받은 것
}을 리턴해주는 함수를 만들어서 넣으면 더 좋을 것 같아요.
그리고 filterdData에 오타가 있는 것 같아요

imhjlov replied:
넵! 수정했습니다.🥰

blingblin-g commented on src/components/filter/FilterInput.tsx:
```diff
@@ -1,11 +1,41 @@
+
+  useEffect(()=>{
+    setCopiedTodos(getTodoStorage());
+    setOriginalTodos(getTodoStorage());
```
localStorage는 파일로 관리되기 때문에 파일에 직접 접근하는 것은 효율이 좋지 못합니다. 접근을 최소화할 수 있도록 변수에 저장해서 쓰는 것이 좋을 것 같아요. state로 관리하면 왜 무한렌더링이 되는지 알아봐야할 것 같네요. 최대한 state를 이용하는 것이 좋을 것 같습니다.

blingblin-g replied:
그리고 이 상태에서 필터링 도중 드롭다운 네임을 바꿔버리면 로컬 스토리지에 있던 기존의 데이터가 다 날아가버리는데, 혹시 의도한 동작인가요? input 창에 입력을 할 때마다 로컬 스토리지에 접근해서 데이터를 바꿔버리는데, 이러면 효율도 문제지만 데이터가 보존이 되지 않습니다. state로 관리하는 방법을 찾아야 할 것 같습니다.

imhjlov replied:
수진님! 리뷰 감사합니다😁 수진님 말대로 localStorage를 변수로 저장해서 쓰는 것으로 수정했습니다.
input창 입력할때 스토리지에 접근하지 않고, state로 관리하도록 바꿔보았습니다!

정말 본인의 코드처럼 잘 아시네요ㅠ 제가 하면서 이 정돈 괜찮겠지 하고 넘어간 부분을 잘 집어주셔서 다시 한 번 고쳐 보는 계기가 되었습니다😍

blingblin-g commented on src/contexts/Todo/TodoStore.tsx:
```diff
@@ -59,6 +60,8 @@ function todosReducer(preState: TodosState, action: Action): TodosState {
       saveTodoStorage(newTodoList);
       return newTodoList;
+    case 'FILTER' : 
+      return action.filterTodo;
```
이 부분은 어디에서 쓰는 건가요?

imhjlov replied:
아 쓰려고 만들었다가 지우는 것을 깜빡했네요😅 감사합니다.

blingblin-g reviewed (changes requested):
고생 많으셨습니다. 다만 기존 데이터가 유실되므로 localStorage가 아니라 state로 관리하는 방법으로 전환해야할 것 같습니다.

dongwonnn committed:
[#8] Style: filterdData 오타 수정

imhjlov committed:
[#8] chore: copy 액션 함수로 todos를 복사

imhjlov committed:
[#8] feat: 특정 todos를 copy할 수 있는 기능

imhjlov committed:
[#8] refactor: context를 통해 filtering되도록 수정

dongwonnn committed:
[#8] refactor: 리뷰 반영

- localStorage 직접 접근을 state로 관리
- 드랍다운 아이템 선택 시 localStorage 비워지는 오류 해결
- include로 filtering 교체

dongwonnn committed:
[#8] fix: conflict 해결
